src/data/rel.lean

```diff
@@ -0,0 +1,187 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Avigad
+
+Operations on set-valued functions, aka partial multifunctions, aka relations.
+-/
+import tactic.ext tactic.interactive data.set.basic order.complete_lattice
+
+variables {α : Type*} {β : Type*} {γ : Type*}
+
+def rel (α : Type*) (β : Type*):= α → β → Prop
+
+namespace rel
+
+variables {δ : Type*} (r : rel α β)
+
+instance : lattice.complete_lattice (rel α β) :=
+by unfold rel; apply_instance
+
+def inv : rel β α := flip r
+
+lemma inv_def (x : α) (y : β) : r.inv y x ↔ r x y := iff.rfl
+
+lemma inv_inv : inv (inv r) = r := by { ext x y, reflexivity }
+
+def dom := {x | ∃ y, r x y}
+
+def codom := {y | ∃ x, r x y}
+
+lemma codom_inv : r.inv.codom = r.dom := by { ext x y, reflexivity }
+
+lemma dom_inv : r.inv.dom = r.codom := by { ext x y, reflexivity}
+
+def comp (r : rel α β) (s : rel β γ) : rel α γ :=
+λ x z, ∃ y, r x y ∧ s y z 
+
+local infixr ` ∘ ` :=rel.comp
+
+lemma comp_assoc (r : rel α β) (s : rel β γ) (t : rel γ δ) :
+  (r ∘ s) ∘ t = r ∘ s ∘ t :=
+begin
+  unfold comp, ext x w, split,
+  { rintros ⟨z, ⟨y, rxy, syz⟩, tzw⟩, exact ⟨y, rxy, z, syz, tzw⟩ },
+  rintros ⟨y, rxy, z, syz, tzw⟩, exact ⟨z, ⟨y, rxy, syz⟩, tzw⟩
+end
+
+@[simp]
+lemma comp_right_id (r : rel α β) : r ∘ @eq β = r :=
+by { unfold comp, ext y, simp }
+
+@[simp]
+lemma comp_left_id (r : rel α β) : @eq α ∘ r = r :=
+by { unfold comp, ext x, simp }
+
+lemma inv_id : inv (@eq α) = @eq α :=
+by { ext x y, split; apply eq.symm }
+
+lemma inv_comp (r : rel α β) (s : rel β γ) : inv (r ∘ s) = inv s ∘ inv r :=
+by { ext x z, simp [comp, inv, flip, and.comm] }
+
+def image (s : set α) : set β := {y | ∃ x ∈ s, r x y}
+
+lemma mem_image (y : β) (s : set α) : y ∈ image r s ↔ ∃ x ∈ s, r x y :=
+iff.refl _
+
+lemma image_mono {s t : set α} (h : s ⊆ t) : r.image s ⊆ r.image t :=
+assume y ⟨x, xs, rxy⟩, ⟨x, h xs, rxy⟩
+
+lemma image_inter (s t : set α) : r.image (s ∩ t) ⊆ r.image s ∩ r.image t :=
+assume y ⟨x, ⟨xs, xt⟩, rxy⟩, ⟨⟨x, xs, rxy⟩, ⟨x, xt, rxy⟩⟩
+
+lemma image_union (s t : set α) : r.image (s ∪ t) = r.image s ∪ r.image t :=
+set.eq_of_subset_of_subset
+  (λ y ⟨x, xst, rxy⟩, 
+    begin 
+      cases xst with xs xt, 
```

Comment:
The anonymous constructor in the lambda allows nesting, so this `cases` can go there.

Reply:
No, this `xst` is an `or`.